### ISSUE TITLE
boolean operator precedence

### ISSUE THREAD

**Issue by crehn** [TNG/ArchUnit#213]
Consider the following scenario:

```java
classes()
    .should(A)
    .andShould(B)
    .orShould(C)
    .andShould(D)
```

Typically boolean `and` is evaluated before boolean `or`. This would result in `(A and B) or (C and D)`. ArchUnit currently has it left associative: `(((A and B) or C) and D)`. I guess it would be quite hard to implement some kind of operator precedence. That's not perfect but it's OK. 

My current rule of thumb is that no and() should follow an or() because this results in a test that is somewhat unintuitive to read and might do something I don't want it to do.

* Is left-associativity part of the contract or is it a detail that may change?
* Could you please document the current behavior and maybe a recommendation on how to mitigate this issue?
* Instead of  providing a rule of thumb (like the one mentioned above) one might think about changing the API such that you cannot add an `and` once there was an `or`. This might be easier to implement than real operator precedence while still ensuring that certain mistakes aren't made.

**Comment by codecholeric** [comment]
Yes, operator precedence would have been quite hard to implement, so I decided to do everything left associative, and yes, it's probably not documented the way it should be :wink: 
This behavior will not change though, because it would break a lot of existing `ArchRules`, and we always try to be as backwards compatible as possible. I guess some Javadoc would be a good idea.

To be honest, I hardly mix `or` and `and`, simply because I think it gets confusing. I rather use the predefined conditions and join them in a way that makes the precedence clear, e.g.:

```
noClasses().should(
        resideInAnyPackage("..foo..").and(
                beAnnotatedWith(Deprecated.class).<JavaClass>forSubType().or(bePublic()))
        )
```

And yes, unfortunately with the limited possibilities to declare types in Java, at some point one needs some ugliness like `<JavaClass>forSubType()` or `ArchConditions.<JavaClass>beAnnotatedWith(..)` :frowning_face: At least AFAIK (if anybody has a hint how to avoid this, I'd be super glad!)

Do you think this approach should be added to the Javadoc of `andShould()` and `orShould()`?
Also, what about `that()..and()..or()`, did you ever get confused by this? I guess the same docs should be added there as well.

**Comment by crehn** [comment]
The pattern I currently use the most is this:

```java

classes() that()...
    .should(exist())
    .andShould()...
    .andShould()...
    .orShould().be(SomeExceptionalCase.class);
    .orShould().be(SomeOtherExceptionalCase.class);
```

where `exist()` is my fix for problem #178:

```java
public static <T> ArchCondition<T> exist() {
    return new ArchCondition<T>("exist") {
        private Set<T> allElements = new HashSet<>();

        @Override
        public void check(T elements, ConditionEvents conditionEvents) {
            allElements.add(elements);
        }

        @Override
        public void finish(ConditionEvents events) {
            boolean conditionSatisfied = allElements.size() > 0;
            String message = String.format("there is/are %d element(s)", allElements.size());
            events.add(new SimpleConditionEvent(allElements.size(), conditionSatisfied, message));
        }
    };
}
```

So there is the existence check, then some conditions to check for (joined with `andShould()`) and in the end there are some exceptional cases I want to exclude. That works quite well. But it took me some time to realize where the existence check had to go in order to work properly. It cannot be put at the end after some `orShould()` for example. That's fine but some hint on that behavior would have helped me.

I like the method chaining style because it gives me better code completion and I don't have to pay attention to parentheses. If the logic is too complicated, I tend to create my own ArchConditions so that the test at hand stays clean and readable. That's why I didn't come across the `forSubType` problem (which is quite ugly to read).

If I ever come across such a situation, I'd rather avoid `forSubType` completely. Even if I understand what's going on and why this is necessary, team members who haven't spent that much time using ArchUnit would certainly be confused or -- even worse -- scared off.

So I'd rather use something like this:

```java
@ArchTest
public static final ArchRule foo = noClasses()
    .should(resideInAnyPackage("..foo..")
        .and(beAnnotatedClassWith(Deprecated.class)
        .or(bePublic()))
);

private static <T> ArchCondition<JavaClass> beAnnotatedClassWith(Class<? extends Annotation> type) {
    return beAnnotatedWith(type);
}
```

Anyway, I think the left-associativity should be explained not only in JavaDoc but also in some section of the user guide. As for `that().and()..or()` I tend to use either really simple expressions or I have a custom Predicate that expresses a concept of my architecture. Nevertheless the left-associativity should be documented there as well.
